src/float.rkt

```diff
@@ -4,61 +4,40 @@
-  (integer-bytes->integer (real->floating-point-bytes x 4) #f))
-
-(define (single-flonum->ordinal x)
-  (cond [(x . < . 0.0f0) (- (single-flonum->bit-field (- 0.0f0 x)))]
-	[else            (single-flonum->bit-field (abs x))]))
-
-(define (single-flonums-between x y)
-  (- (single-flonum->ordinal y) (single-flonum->ordinal x)))
+(provide midpoint ulp-difference *bit-width* ulps->bits bit-difference
+         sample-double </total <=/total =-or-nan? nan?-all-types ordinary-value?)
 
 (define (ulp-difference x y)
-  (match* (x y)
-    [((? real?) (? real?))
-     (if (flag-set? 'precision 'double) (flonums-between x y) (single-flonums-between x y))]
-    [((? complex?) (? complex?))
-     (+ (ulp-difference (real-part x) (real-part y))
-        (ulp-difference (imag-part x) (imag-part y)))]
-    [((? boolean?) (? boolean?))
-     (if (equal? x y) 0 64)]
-    [((? posit8?) (? posit8?))
-     (p8s-between x y)]
-    [((? posit16?) (? posit16?))
-     (p16s-between x y)]
-    [((? posit32?) (? posit32?))
-     (p32s-between x y)]
-    [((? quire8?) (? quire8?))
-     (p8s-between (quire8->posit8 x) (quire8->posit8 y)) ]
-    [((? quire16?) (? quire16?))
-     (p16s-between (quire16->posit16 x) (quire16->posit16 y))]
-    [((? quire32?) (? quire32?))
-     (p32s-between (quire32->posit32 x) (quire32->posit32 y))]))
+  (if (and (complex? x) (complex? y) (not (real? x)) (not (real? y)))
+    (+ (ulp-difference (real-part x) (real-part y))
+       (ulp-difference (imag-part x) (imag-part y)))
+    (let ([->ordinal (representation-repr->ordinal (match* (x y)
+        [((? real?) (? real?)) (if (flag-set? 'precision 'double) binary64 binary32)]
+        [((? boolean?) (? boolean?)) bool]
+        [((? posit8?) (? posit8?)) posit8]
+        [((? posit16?) (? posit16?)) posit16]
+        [((? posit32?) (? posit32?)) posit32]
+        [((? quire8?) (? quire8?)) quire8]
+        [((? quire16?) (? quire16?)) quire16]
+        [((? quire32?) (? quire32?)) quire32]))])
+      (- (->ordinal y) (->ordinal x)))))
 
-(define (midpoint-float p1 p2)
-  (cond 
-   [(and (double-flonum? p1) (double-flonum? p2))
-    (flstep p1 (quotient (flonums-between p1 p2) 2))]
-   [(and (single-flonum? p1) (single-flonum? p2))
-    (floating-point-bytes->real
-     (integer->integer-bytes
-      (quotient
-       (+ (single-flonum->ordinal p1) (single-flonum->ordinal p2))
-       2)
-      4 #f) #f)]
-   [(and (posit8? p1) (posit8? p2))
-    ;; NOTE: This isn't a binary search (just splits the difference)
-    (posit8-div (posit8-add p1 p2) (double->posit8 2.0))]
-   [(and (posit16? p1) (posit16? p2))
-    (posit16-div (posit16-add p1 p2) (double->posit16 2.0))]
-   [(and (posit32? p1) (posit32? p2))
-    (posit32-div (posit32-add p1 p2) (double->posit32 2.0))]
-   [else
-    (error "Mixed precisions in binary search")]))
+;; Returns the midpoint of the representation's ordinal values,
+;; not the real-valued midpoint
+(define (midpoint p1 p2)
+  (define repr (cond
```

Comment:
I think we should pull "repr interrogation" into its own function, though I could live with just eliminating it as part of the next phase.

Reply:
Yeah, I think part of the next phase is going to be passing around the repr context so that we don't have to do this interrogation. Definitely a rough edge right now though.